Allow passing style to `Icon` for mobile

pinarol opened the pull request:
## Description
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/1004

gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1017

## How has this been tested?

Checked icons both on web and mobile side and verified that they are rendered properly.

More detail for mobile testing can be found at gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1017

## Types of changes

Passing `className` as a prop doesn't work for mobile, react-native components instead expect a style object. To be able to change the color of the Icon we've added the `style` prop. But this time we are introducing a new prop which is not useful for web. So it might not be the ideal solution. If there's a better way of doing this please let me know. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

pinarol committed:
Allow passing style to `Icon` for mobile

Tug commented on packages/components/src/icon/index.js:
```diff
@@ -8,24 +8,24 @@ import { cloneElement, createElement, Component, isValidElement } from '@wordpre
 		}
 
-		return icon();
+		return icon( { style } );
```
In theory we should pass all the props here instead.

pinarol replied:
I thought the same but it didn't have the other props before so didn't want to cause regression

Tug replied:
Updated with all the props!
I looked over the codebase and could not see a case where this change would break things

Tug committed:
Use the Icon component in image and video blocks to style our svgs

pinarol commented:
Update: with https://github.com/WordPress/gutenberg/pull/15778/commits/06949b6ce9eeb265359bf42bd40b4d4435809435 we are refactoring image/video blocks on mobile side to use the updated `Icon` component. I've tested those and they work OK @Tug

pinarol commented on packages/components/src/icon/index.js:
```diff
@@ -8,24 +8,24 @@ import { cloneElement, createElement, Component, isValidElement } from '@wordpre
  */
```
Hi @tyxla @youknowriad 👋 As far as I can see you are the authors of `Icon` component. Could you be able to review our changes? Thanks

tyxla replied:
Hey @pinarol! I'm no longer working on Gutenberg for a while now, so I'd defer to someone from @WordPress/gutenberg-core to review this one.

aduth commented on packages/components/src/icon/index.js:
```diff
@@ -34,6 +34,7 @@ function Icon( { icon = null, size, className } ) {
 			width: iconSize,
 			height: iconSize,
 			...icon.props,
+			...style,
```
Is this meant to be `...` spread into the _top-level_ props?

Tug replied:
This is on purpose. The idea here is to be able to overwrite any properties using the following construct:
```jsx
<Icon icon={ svgIcon } style={ newSvgStyles } />;
```
That includes `width` and `height` that we'd like to set to `"100%"` for instance so they fit our parent container. Or the `fill` property that could already be present on the original svg and that we might want to overwrite.

Same for `Dashicon`, we need to spead `style` so they are transformed into `extraProps` here: https://github.com/WordPress/gutenberg/blob/master/packages/components/src/dashicon/index.js#L907

aduth replied:
> This is on purpose. The idea here is to be able to overwrite any `SVG` attributes using the following construct:

My initial reaction here is that we're conflating style properties with DOM attributes, and that there isn't a guaranteed overlap, just like we can't have `<div background-image="..." />`.

But then I found this interesting mention in the specification, which seems specific to SVG:

>For each styling property defined in this specification (see Property Index), there is a corresponding XML attribute (the presentation attribute) with the same name that is available on all relevant SVG elements.

https://www.w3.org/TR/SVG11/styling.html#UsingPresentationAttributes

So I guess it _is_ fair to assume any style property can be assigned as a DOM attribute.

It still seems misleading to accept this as a named `style` object, only then to proceed to apply it as spread on the rendered element, rather than to just pass through all additional props from the `Icon` component (e.g. `...additionalProps` pattern) to the rendered SVG.

That being said, since I was apparently under-informed about how style properties map to DOM attributes for SVG, maybe this is an expected behavior.

pinarol replied:
@aduth I think I got your point on this, I agree, having a prop just to use for spreading at the same time naming it as `style` doesn't make that much sense on this level. I have changed the code, would you mind giving it another look?

I think we got to that point because we are passing objects from our `.native.scss` files as a value to it. But actually the `style` here doesn't correspond to the `style` in react-native, so it doesn't need to have that name.

koke commented:
From the comments in https://github.com/WordPress/gutenberg/pull/15863#discussion_r288081056, let's make sure we revert all the icon.js files to their previous state and export `SVG` elements by default, and not functions.

pinarol committed:
Update style as additionalProps

pinarol committed:
Merge branch 'master' into rnmobile/add-style-to-icon

pinarol committed:
Fix syntax

pinarol committed:
Revert README

Tug commented on packages/components/src/icon/index.js:
```diff
@@ -8,24 +8,24 @@ import { cloneElement, createElement, Component, isValidElement } from '@wordpre
 	if ( 'function' === typeof icon ) {
 		if ( icon.prototype instanceof Component ) {
-			return createElement( icon, { className, size: iconSize } );
+			return createElement( icon, { className, additionalProps, size: iconSize } );
```
We should be spreading `additionalProps` here, it's not a props name

pinarol replied:
Done

Tug commented on packages/components/src/icon/index.js:
```diff
@@ -8,24 +8,24 @@ import { cloneElement, createElement, Component, isValidElement } from '@wordpre
 		}
 
-		return icon();
+		return icon( { className, additionalProps, size: iconSize } );
```
Same here:
```js
return icon( { className, ...additionalProps, size: iconSize } );
```

pinarol replied:
Done

Tug commented on packages/components/src/icon/index.js:
```diff
@@ -42,6 +43,7 @@ function Icon( { icon = null, size, className } ) {
 	if ( isValidElement( icon ) ) {
 		return cloneElement( icon, {
 			className,
+			additionalProps,
```
And here :)

pinarol replied:
Done

Tug commented:
Thanks for taking over here @pinarol, couldn't focus on this one as much as I'd like to.
I reviewed your changes and I think you forgot to spread additionalProps at a few places, other than that it looks good to me :)

pinarol committed:
Spread additionalProps

Tug reviewed (approved):
LGTM 👍

aduth commented on packages/components/src/icon/index.js:
```diff
@@ -8,24 +8,24 @@ import { cloneElement, createElement, Component, isValidElement } from '@wordpre
 import { Dashicon, SVG } from '../';
 
-function Icon( { icon = null, size, className } ) {
+function Icon( { icon = null, size, className, ...additionalProps } ) {
```
Technically `className` is redundant, since it would be applied now with the behavior of `additionalProps` anyways.

aduth committed:
Components: Icon: Update documentation for props pass-through

aduth committed:
Compoonents: Icon: Update unit tests for props pass-through

aduth committed:
Components: Icon: Remove redundant explicit className handling

aduth committed:
Components: Icon: Apply additionalProps consistently as last of props

There should be no effective difference with `size` being an explicitly named prop not overridable via additionalProps

aduth reviewed (approved):
From the web side, this looks good. I've pushed a few changes directly to the branch, since these changes should be reflected both in documentation and in unit tests. The unit tests were refactored to better reflect behaviors of the Icon component we're expecting to be common amongst the various possible logic paths, so we're more conscious of consistency (not a reflection on this pull request, but a general point on implementation of these sorts of multi-return components).

pinarol commented:
Thank you very much for the contribution & review @aduth !

pinarol committed:
Merge branch 'master' into rnmobile/add-style-to-icon
